discovery/gossiper.go

```diff
@@ -2425,91 +2292,48 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
-		var b bytes.Buffer
-		if err := msg.Encode(&b, 0); err != nil {
-			return err
+		// If the channel cannot be found, it is most likely a leftover
+		// message for a channel that was closed, so we can consider it
+		// stale.
+		if err == channeldb.ErrEdgeNotFound {
+			return true
+		}
+		if err != nil {
+			log.Debugf("Unable to retrieve channel=%v from graph: "+
+				"%v", err)
+			return false
 		}
 
-		// Add the encoded message to the database using the peer
-		// + shortChanID as key.
-		return bucket.Put(key[:], b.Bytes())
-
-	})
-	if err != nil {
-		return err
-	}
-
-	// We have succeeded adding the message to the database. We now launch
-	// a goroutine that will keep on trying sending the message to the
-	// remote peer until it succeeds, or the gossiper shuts down. In case
-	// of success, the message will be removed from the database.
-	d.wg.Add(1)
-	go func() {
-		defer d.wg.Done()
-		for {
-			log.Debugf("Sending AnnounceSignatures for channel "+
-				"%v to remote peer %x", msg.ChannelID,
-				remotePeer.SerializeCompressed())
-			err := d.cfg.SendToPeer(remotePeer, msg)
-			if err == nil {
-				// Sending succeeded, we can
-				// continue the flow.
-				break
-			}
-
-			log.Errorf("unable to send AnnounceSignatures message "+
-				"to peer(%x): %v. Will retry when online.",
-				remotePeer.SerializeCompressed(), err)
-
-			peerChan := make(chan lnpeer.Peer, 1)
-			d.cfg.NotifyWhenOnline(remotePeer, peerChan)
-
-			select {
-			case <-peerChan:
-				// Retry sending.
-				log.Infof("Peer %x reconnected. Retry sending"+
-					" AnnounceSignatures.",
-					remotePeer.SerializeCompressed())
+		// If the proof exists in the graph, then we have successfully
+		// received the remote proof and assembled the full proof, so we
+		// can safely delete the local proof from the database.
+		return chanInfo.AuthProof != nil
 
-			case <-d.quit:
-				log.Infof("Gossiper shutting down, did not " +
-					"send AnnounceSignatures.")
-				return
-			}
+	case *lnwire.ChannelUpdate:
+		// The MessageStore will always store the latest ChannelUpdate
+		// as it is not aware of its timestamp (by design), so it will
+		// never be stale. We should still however check if the channel
+		// is part of our graph. If it's not, we can mark it as stale.
+		_, _, _, err := d.cfg.Router.GetChannelByID(msg.ShortChannelID)
+		if err != nil && err != channeldb.ErrEdgeNotFound {
+			log.Debugf("Unable to retrieve channel=%v from graph: "+
+				"%v", err)
 		}
+		return err == channeldb.ErrEdgeNotFound
 
-		log.Infof("Sent channel announcement proof to remote peer: %x",
-			remotePeer.SerializeCompressed())
-	}()
-
-	// This method returns after the message has been added to the database,
-	// such that the caller don't have to wait until the message is actually
-	// delivered, but can be assured that it will be delivered eventually
-	// when this method returns.
-	return nil
+	default:
+		// Unsupported message type.
+		return true
```

Comment:
idt this is safe bc of the concurrency issue @halseth mentioned. if the new type requires comparing timestamps for example, this would cause them to be removed if a new message is added in the meantime. The safer fix seems to be creating a helper around `MessagesForPeer` that filters out unknown message types

Reply:
Thought we had agreed that this could be deferred if and when a new message type is introduced?

Reply:
i think the requirement then would be we have to do a migration after every new message is added, unless this logic also changes to not delete unknown message types in the future

Reply:
seems safer just to filter and ignore unknown things. along the lines of "i probably shouldn't delete it if i don't know what it is"

Reply:
> unless this logic also changes to not delete unknown message types in the future

Yep that's what I'm referring to. Not worth the trouble IMO, we know we won't add any other message types in the near future, and #1595 will be introducing a migration for 0.6.

Reply:
> we know we won't add any other message types in the near future

we were discussing this possibility the other day, wrt to channel close message. idk it's that unlikely that this could be extended to cover more messages. IMO it's not worth having to remember the extra constraints on adding message+migration if we can address it here and forget about it. just my opinion though

Reply:
Fixed.